### ISSUE TITLE
Can the XR___LayerFactory names be more general purpose?

### ISSUE THREAD

**Issue by toji** [immersive-web/layers#38]
I've suggested this before, but felt at the time it was more important to get some of the basic usage patterns nailed down first. Now that we've established the broad strokes of the API and I've got a concrete use case outside of layers to point at as an example I'd like to revisit it.

In the Lighting Estimation API we need to deliver a cube map in an API-appropriate texture format. In order to make the WebXR/WebGL/(Eventual WebGPU) interaction easier the [current explainer IDL](https://github.com/immersive-web/lighting-estimation/blob/master/lighting-estimation-explainer.md#appendix-a-proposed-partial-idl) proposes using the `XRWebGLLayerFactory` as the method of surfacing that texture, since it's already well positioned to serve up WebGL-specific resources for the session. 

The only real issue with this route is that it makes the name `XRWebGLLayerFactory` a misnomer, since it would provide more than just layers. [@kearwood and I would both prefer](https://github.com/immersive-web/lighting-estimation/pull/26#issuecomment-604656806) to see the name changes to something more generic as a result.

I'm really flexible on what that name is, as long as it suggests that it handles more WebGL resources than just layers. A few suggestions in no particular order of preference.

- `XRWebGLInterface` (may be confused with the IDL "interface" concept)
- `XRWebGLBinding`
- `XRWebGLGraphicsBinding`
- `XRWebGLSessionBinding`
- `XRWebGLResourceProvider`

**Comment by cabanier** [comment]
We could drop the "layer" part and rename it to `XRWebGLFactory`.
Otherwise, I have a very slight preference to `XRWebGLInterface` and `XRWebGLBinding `

**Comment by RafaelCintron** [comment]
Reading through the layers proposal, I am having a hard time understanding why we need the concept of "factories" or "bindings" in the first place.  

Instead of:

````JS
const gl = canvas.getContext('webgl', { xrCompatible: true });
const glLayerFactory = new XRWebGLLayerFactory(xrSession, gl);
const layer = await glLayerFactory.requestProjectionLayer(gl.TEXTURE_2D, { alpha: false });
````

why can't we have?

````JS
const gl = canvas.getContext('webgl', { xrCompatible: true });
const layer = await xrSession.requestWebGLProjectionLayer(gl, gl.TEXTURE_2D, { alpha: false });
````

The second version is less code and still feature detectable by web developers. 

Having a separate session specific binding object seems like it might add confusion about which object instances you need to get a hold of to do your job, especially if you're creating multiple sessions.  

What happens if you make more than one binding object per session of the same or different graphics API?  If one of the binding objects gets garbage collected, has the developer lost any state they can't get back?

**Comment by cabanier** [comment]
> Having a separate session specific binding object seems like it might add confusion about which object instances you need to get a hold of to do your job, especially if you're creating multiple sessions.

@RafaelCintron We talked about that during a weekly call but. unfrotunately I can't find the notes.
If I remember correctly, @toji said that this approach is cleaner because otherwise we'd have to spec that the session object has conditional interfaces which is an undesired pattern.

> What happens if you make more than one binding object per session of the same or different graphics API? 

I had filed that as an issue in the spec. For now, yes, you're allowed to create more than one object. 

> If one of the binding objects gets garbage collected, has the developer lost any state they can't get back?

No. As you correctly observed, the factory is just a convenience method. As long as the session is active, its associated layers will be valid. Likewise, as long as the XRFrame is active, subimamge are valid.

**Comment by RafaelCintron** [comment]
> @RafaelCintron We talked about that during a weekly call but unfortunately I can't find the notes.

@cabanier I do not regularly attend the weekly immersive web calls but I do not remember this issue being on the agenda.  Is there a layers-specific weekly call?  If so, where can I find more information?

**Comment by cabanier** [comment]
> @cabanier I do not regularly attend the weekly immersive web calls but I do not remember this issue being on the agenda. Is there a layers-specific weekly call? If so, where can I find more information?

No. There is no layers specific call. We've covered layers topics during the regular weekly call.

**Comment by toji** [comment]
@rik has already covered it pretty well, but I wanted to chime in that the primary intent of making a separate interface for this is API clarity and separation of concerns. It's certainly true that the layer mechanism could live on the session, but given what we expect to support over time it would grow pretty unwieldy. Consider this:

```webidl
partial interface XRSession {
  Promise<XRProjectionLayer> requestWebGLProjectionLayer(WebGLRenderingContext, GLEnum, InitDictionary);
  Promise<XRQuadLayer> requestWebGLQuadLayer(WebGLRenderingContext, GLEnum, InitDictionary);
  Promise<XRCylinderLayer> requestWebGLCylinderLayer(WebGLRenderingContext, GLEnum, InitDictionary);
  Promise<XREquirectLayer> requestWebGLEquirectLayer(WebGLRenderingContext, GLEnum, InitDictionary);
  Promise<XRCubeLayer> requestWebGLCubeLayer(WebGLRenderingContext, GLEnum, InitDictionary);

  XRWebGLSubImage? getWebGLSubImage(XRLayer, XRFrame);
  XRWebGLSubImage? getWebGLViewSubImage(XRLayer, XRView);

  WebGLTexture? getReflectionCubeMap(WebGLRenderingContext, XRLightProbe);
  // Any other functions that need WebGL resources.

  Promise<XRProjectionLayer> requestWebGPUProjectionLayer(GPUDevice, InitDictionary);
  Promise<XRQuadLayer> requestWebGPUQuadLayer(GPUDevice, InitDictionary);
  Promise<XRCylinderLayer> requestWebGPUCylinderLayer(GPUDevice, InitDictionary);
  Promise<XREquirectLayer> requestWebGPUEquirectLayer(GPUDevice, InitDictionary);
  Promise<XRCubeLayer> requestWebGPUCubeLayer(GPUDevice, InitDictionary);

  XRWebGPUSubImage? getWebGPUSubImage(XRLayer, XRFrame);
  XRWebGPUSubImage? getWebGPUViewSubImage(XRLayer, XRView);

  GPUTextureView? getReflectionCubeMap(GPUDevice, XRLightProbe);
  // Any other functions that need WebGPU resources.

  Promise<XRQuadLayer> requestVideoQuadLayer(HTMLVideoElement, InitDictionary);
  Promise<XRCylinderLayer> requestVideoCylinderLayer(HTMLVideoElement, InitDictionary);
  Promise<XREquirectLayer> requestVideoEquirectLayer(HTMLVideoElement, InitDictionary);

  // The same set for ImageBitmaps eventually?

  // At some point DOM elements are gonna be part of this as well...
}
```

Considering the `XRSession` originally had 5 methods on it, this represents a 4X increase in the complexity of our core interface. I'd personally rather have these methods be well grouped, both as an API user and implementer.

There is also some benefits that come from being able to "pair" the graphics API and the session only once. Some validation that would need to happen with every call can be avoided, for example. (Though it's admittedly small.) There's also some cases like with the reflection cube maps where the separate interface allows for cleaner re-use of resources like textures, since you don't have to track an internal map of contexts->textures in the session.

**Comment by rik** [comment]
@toji I'm probably not the Rik you're trying to mention :)

**Comment by toji** [comment]
Oops! Apologies to both @rik and @cabanier (AKA: The "Right" Rik) for the mixup!

**Comment by toji** [comment]
Responding to @cabanier's earlier suggestion, I don't mind `XRWebGLFactory`, with the exception of the fact that it sounds ever so slightly like it's creating WebGL contexts? I doubt that would be a problem long term, though.

In any case, of the options you listed I'd lean towards `XRWebGLInterface` or `XRWebGLBinding`, probably with a slight preference for the  `Binding` verbiage simply because it matches the [OpenXR name for the same concept.](https://www.khronos.org/registry/OpenXR/specs/1.0/man/html/openxr.html#_xrgraphicsbindingvulkankhr3)

**Comment by cabanier** [comment]
We can do a poll in the W3C slack :-)

**Comment by RafaelCintron** [comment]
@toji and @cabanier, I personally do not mind additional methods on the session object but your arguments make sense. 

Though, if we continue with the pattern of "new XR module == new factory object", the web developer will need to manage an increasing number of silo-ed objects (all tied to a session) that must be passed to each other as parameters at the right times in order for things to happen.  At the implementation level, we'll need to handle cases where objects created from multiple instances of the same silo are used with one another and decide whether that's an error or not.  If everything comes from a session, on the other hand, it seems more clear how things fit together.

**Comment by cabanier** [comment]
@RafaelCintron do you want to open a new issue on the need for separate factory method and put it on the agenda?

**Comment by toji** [comment]
@RafaelCintron: I want to avoid the need for some new factory object per module, which is part of the reason I'm looking to use this particular factory for things like surfacing lighting estimation cube maps.

**Comment by RafaelCintron** [comment]
> I want to avoid the need for some new factory object per module, which is part of the reason I'm looking to use this particular factory for things like surfacing lighting estimation cube maps.

@toji , fair enough. As part of specing the factory, we should consider lifetimes and what happens when developers make multiple factories and inter-mix objects between them. If the factory is purely meant as an object creator, then I'm not worried. If the factory allows state mutation on it, then we'll want to thing about the rules around inter-mixing and how much we allow it.

**Comment by cabanier** [comment]
> @toji , fair enough. As part of specing the factory, we should consider lifetimes and what happens when developers make multiple factories and inter-mix objects between them. If the factory is purely meant as an object creator, then I'm not worried. If the factory allows state mutation on it, then we'll want to thing about the rules around inter-mixing and how much we allow it.

I will make sure that there's nothing in the spec that states that the factories change state. I will also add a note that multiple factories are allowed.

**Comment by cabanier** [comment]
@RafaelCintron Please see #47 
I tried to make you a reviewer but for some reason you don't show up in the list.

**Comment by RafaelCintron** [comment]
@cabanier I think I do not show up because I haven't been made a "collaborator" for the layers repository.  Not sure how I can change that.

**Comment by cabanier** [comment]
I see that you're a reviewer now :-)

**Comment by cabanier** [comment]
@toji should we have a poll to decide on the name? I can also put it on the agenda to see how people feel.

**Comment by toji** [comment]
/agenda to ask about XR/WebGL interop API naming.

We've missed this week's email cutoff, but given that we're discussing a lot of layer related things this time around anyway I think we can squeeze it in tomorrow.

**Comment by toji** [comment]
Straw poll on today's call indicated a strong preference for "XRWebGLBinding"

**Comment by toji** [comment]
Closed by #53
